cas/src/main/java/org/springframework/security/cas/web/CasAuthenticationFilter.java

```diff
@@ -238,8 +244,14 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
 		}
 		String serviceTicket = obtainArtifact(request);
 		if (serviceTicket == null) {
-			this.logger.debug("Failed to obtain an artifact (cas ticket)");
-			serviceTicket = "";
+			if (this.gateway) {
+				this.logger.debug("Gateway request with no CAS ticket");
+				return new CasGatewayNoServiceTicketAuthenticationToken();
```

Comment:
What happens on a public URL when there is no `AuthenticationException`? Will the `CasAuthenticationEntryPoint` be invoked?
To be able to invoke this filter to different URLs, the `filterProcessingUrl` should allow customization, however, I'm still not convinced that it is the responsibility of this filter to be aware of the gateway request.

If we take a look at how OAuth2 components work, there are at least two filters, one that performs the redirect to the Authorization Server and another that processes the authorization response.

Reply:
The `CasAuthenticationEntryPoint` will only be invoked if the URL is configured with `authenticated`.

And the `CasGatewayNoServiceTicketAuthenticationToken` cannot be an `AnonymousAuthenticationToken` as it leads to an infinite loop.

On the other hand, this prevents triggering a new CAS round-trip and only leads to a 403 error page for another URL protected with non-gateway CAS login. It would certainly need a custom `AccessDeniedHandler`.

So although re-using the `CasAuthenticationFilter` for the `gateway` setting seems an alluring idea, I think I was wrong on this: I will create a new filter. I will do that next Monday.

I think it's better to close this PR and submit a new PR, isn't it?

Reply:
OK. I will do that today. Thanks for your time and advices

Reply:
@marcusdacoregio I just submitted a polished-to-main-branch PR from #9881.

Just let me know if you have additional comments.